power: stm32l4: support of the Low Power Mode

FRASTM opened the pull request:
This patch enables the low power Mode for the stm32l4r5 soc. 
The SLEEP (1,2,3) Modes, corresponding to the stm32L4 STOP 0,1,2, Modes, are supported on this device.
The CONFIG_SYS_POWER_MANAGEMENT is required to activate the low power sleep modes.
This config will enable the LPTIM low power timer to preserve the kernel count  during the low power phase. The lptim is a low power timer running during stop Modes, so that the kernel ticks are preserved. Then will the kernel run in a tickless mode, going to low power sleep mode when the device policy accepts. The device PM residency policy applies. 
Due to the LPTIM clocking and timeout capacity, the maximum timeout value is 2000ms : LPTIM counter reaches the 16 bits value, when prescaler is set to 1 (DIV1). This means that when kernel request a sleep duration of more than 2 seconds, it will sleep for 2000ms then wakes Up and goes to sleep again for the next 2-second period. 
Increasing this timeout value implies that the LPTIM count unit is also increased. The LPTIM count unit gives the minimum timeout value. Based on the LSE source clock, the LPTIM count unit is 0.03 ms.
This PR also requires a definition of the PM residency policy. Then any application calling the k_sleep function will enter low power sleep mode for the given duration. Some peripherals can run during stop mode, some cannot. 
This is adjustable with CONFIG_HAS_SYS_POWER_STATE_SLEEP_X.
 
FOREVER: the system goes to sleep without a timer to wakeup, only a reset or EXTI can resume, if the EXTI line has been configured before entering the sleep mode. EXTI line can be a gpio pin or a peripheral with wakeup capability. Then is the lptim able to re-start counting.

Note that DEEP_SLEEP modes are not enabled yet, as LPTIM cannot run during stm32 standby or shutdown modes.

Signed-off-by: Francois Ramu francois.ramu@st.com

pizi-nordic commented on samples/subsys/power/stm32_pm/lowpower/Kconfig:
```diff
@@ -0,0 +1,19 @@
+	bool
+	default y
+	select SYS_POWER_SLEEP_STATES
+	select HAS_SYS_POWER_STATE_SLEEP_1
```
These are properties of SoC, so should be selected by the SoC entry.

FRASTM replied:
--> Soc Kconfig gets the  HAS_SYS_POWER_STATE_SLEEP_xxx
--> sample appli Kconfig gets the SYS_POWER_SLEEP_STATES and SYS_POWER_DEEP_SLEEP_STATES

pizi-nordic commented on samples/subsys/power/stm32_pm/lowpower/src/main.c:
```diff
@@ -0,0 +1,204 @@
+
+	switch (state) {
+
+	case SYS_POWER_STATE_SLEEP_1:
```
Each state could be independently enabled/disabled, so you need #if statements around each case.

FRASTM replied:
OK

erwango commented on subsys/power/device.c:
```diff
@@ -32,6 +32,17 @@ static const char core_devices[NUM_CORE_DEVICES][MAX_DEV_NAME_LEN] = {
 	"sys_clock",
 	"UART_0",
 };
+#elif defined(CONFIG_SOC_SERIES_STM32L4X) \
```
Let's stick to L4 only for now

FRASTM replied:
ok, 
#elif defined(CONFIG_SOC_SERIES_STM32L4X)

erwango commented on samples/subsys/power/stm32_pm/lowpower/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2019 STMicroelectronics
```
I wonder if a prj.conf would be enough here

FRASTM replied:
yes, it works without KConfig

erwango commented on samples/subsys/power/stm32_pm/lowpower/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(l4r5_power_mgr)
```
let it be `stm32_power_mgr`

FRASTM replied:
OK

erwango commented on samples/subsys/power/stm32_pm/lowpower/src/main.c:
```diff
@@ -0,0 +1,232 @@
+
+
+/* Application main Thread */
+void main(void)
```
A README would be nice to allow to undertsand quickly what the sample is supposed to do

FRASTM replied:
comment added to the main.c

FRASTM commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -0,0 +1,150 @@
+{
+	switch (state) {
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
```
soc/arm/nordic_nrf/nrf52/power.c did, though

FRASTM replied:
un-indented

FRASTM commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -0,0 +1,150 @@
+#endif /* CONFIG_SYS_POWER_SLEEP_STATES */
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP_STATES
+ #ifdef CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_1
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
```
after deepsleep low power mode (standby or shutdown) code never comes to _sys_pm_power_state_exit_post_ops, as the system reboots when exiting the low power mode
So in the _sys_pm_power_state_exit_post_ops function, 
case SYS_POWER_STATE_DEEP_SLEEP_1 or 
case SYS_POWER_STATE_DEEP_SLEEP_2 
is never reached

FRASTM commented:
The sample application is using the user button to exit from Deep sleep mode

FRASTM commented on soc/arm/st_stm32/Kconfig.defconfig:
```diff
@@ -1,3 +1,12 @@
+# Kconfig - ST Microelectronics STM32 MCU line
+#
+# Copyright (c) 2016 Open-RnD Sp. z o.o.
```
in soc/arm/st_stm32/stm32l4 and other series, the Kconfig.defconfig.series has the following line:
# Kconfig - ST Microelectronics STM32L4 MCU line

FRASTM replied:
adding # Copyright (c) 2019 STMicroelectronics.

FRASTM commented on soc/arm/st_stm32/Kconfig.defconfig:
```diff
@@ -1,3 +1,12 @@
 
 source "soc/arm/st_stm32/*/Kconfig.defconfig.series"
+
+config SYS_POWER_MANAGEMENT
```
ok

FRASTM commented on soc/arm/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx:
```diff
@@ -31,4 +31,19 @@ config GPIO_STM32_PORTG
 
 endif # GPIO_STM32
 
+config HAS_SYS_POWER_STATE_SLEEP_1
```
removed here

FRASTM commented:
low power  mode sleep and deep sleep, with application example exiting deep sleep mode with user push button

FRASTM commented:
changed copyright in soc/arm/st_stm32/Kconfig.defconfig

FRASTM commented:
rebased 
this PR is for lowpower on L4
 requires  Implementation on L4 series (#18582)
and commit 3617023377d078c6d84fe335f8e316aec0e0b21b

jdascenzio commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -0,0 +1,168 @@
+	 * when OS started idling code.
+	 */
+	irq_unlock(0);
+}
```
Do you still work on this pull request?
I take your code for my project. I restart the clock on wakeup like this
```
struct device* dev = device_get_binding(STM32_CLOCK_CONTROL_NAME);
SystemCoreClock = 160000000;
dev->config->init(dev);
```

aurel32 commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -0,0 +1,168 @@
+		LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+		/* ensure HSI is the wake-up system clock */
+		LL_RCC_SetClkAfterWakeFromStop(LL_RCC_STOP_WAKEUPCLOCK_HSI);
+		LL_PWR_DisableSRAM3Retention();
```
Note that not all L4x SOCs have SRAM3, so this doesn't compile for example with an L432 SoC.

FRASTM replied:
OK, I will add the flag PWR_CR1_RRSTP for support of SRAM3

FRASTM commented:
rebase
control of LL_PWR_DisableSRAM3Retention present or not

FRASTM commented:
fixing conflict in subsys/power/power.c

aurel32 commented:
> fixing conflict in subsys/power/power.c

@FRASTM, it seems the commit "soc: arm: st_stm32 add low power to stm32l4 series" has been lost in the rebase.

aurel32 commented:
I have been playing with that PR, or rather the "soc: arm: st_stm32 add low power to stm32l4 series" commit for a few days on an STM32L422 based board. I haven't use the lptim related committed, and didn't notice any issue related with the timer. Note however that I have configured `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32768` and `CONFIG_STM32_LPTIM_CLOCK=32768`.

Overall I have been able to get it working, with a few limitations though:
- The I2C device doesn't work anymore when `CONFIG_I2C_STM32_INTERRUPT=y`. The wait for the interrupt is done with K_FOREVER which causes the MCU to go into `SYS_POWER_STATE_SLEEP_3` / STOP2. The I2C device is not supported in that mode (at least for I2C1, it seems is not true for I2C3). I have workarounded that by wrapping the `i2c_transfer` calls with `sys_pm_ctrl_disable_state(SYS_POWER_STATE_SLEEP_3)` and `sys_pm_ctrl_enable_state(SYS_POWER_STATE_SLEEP_3)`. Strangely enough, I do not have the issue with the SPI device.
- I am using HSI16 as the CPU clock, as it is the one configured after getting out of sleep. I understand that adding PLL support is complex, but it might be nice to also have MSI support as it the MCU support that directly. MSI allows higher clock speeds (although some changes are needed to the clock driver to allow that) and can work in PLL mode with the LSE.
- I measured a power consumption of around 600µA which is higher than expected. I have other devices on the board that are supposed to be in low power mode and use less than 10µA. I have a spare PCB, I'll solder only the STM32L422 in the next days to check if the issue comes from the MCU or not.

Also I haven't tested deep sleep, but from what I understand from the comments, some features are missing so it might be a good idea to move it to a separate PR.

With all that said, in order to progress on STM32L4 low power support, it might be a good idea to merge that feature without deep sleep and without making `SYS_POWER_MANAGEMENT=y` being the default.

aurel32 commented:
> I measured a power consumption of around 600µA which is higher than expected. I have other devices on the board that are supposed to be in low power mode and use less than 10µA. I have a spare PCB, I'll solder only the STM32L422 in the next days to check if the issue comes from the MCU or not.

I found the issue, the problem is that in STOP mode, the SPI peripheral is disabled, so pins that do not have a pull-down or a pull-up are floating. The MOSI pin is held by a pull-down, but it's not the case for the SCK pin. This causes the SPI slaves to see transitions, increasing the power consumption.

FRASTM commented:
rebase
adding commit `soc: arm: st_stm32 add low power to stm32l4  series`

FRASTM commented:
Changed to set the config SYS_POWER_MANAGEMENT  for this L4r5 only.
DEEP Sleep (standy or shutdown) requires a mechanism to exit : switch user button (gpio PC13) is suitable when configured as a wakeUp pin 2  (or RTC auto-wakeup feature which is not implemented yet). 
Note that exiting from standby or shutdown comes through a system **reset**, so that the function   `_sys_pm_power_state_exit_post_ops()` is not called in the power.c. Then is the PR 'drivers: hwinfo: Add reset cause support' #24884 

useful to get info on the origin of reset.

FRASTM commented:
Changed the low power management for stm32l4. For Sleep Mode (stm32 stop modes) the low power timer lptim1 starts counting when going to sleep. The lptim irq wakes up the kernel.
The lptim is initialsed when enabling the low power and start counting when calling the sys_pm_policy_next_state for a nb of kernel ticks.
With this **fixed** configuration, the lptim is clocked by the LSE with a prescaler of 8, so that possible timeout in ms  is > PM residency and < 16000.

FRASTM commented:
Closing this PR as another approach is now selected with https://github.com/zephyrproject-rtos/zephyr/pull/25628/

FRASTM committed:
soc: arm: st_stm32 add low power to stm32l4  series

This patch introduces the support of low power modes
for the STM32L4xx from STMicroelectronics based on the lptim
Here, the power modes are sleep modes with lptimer as wakeup.
Depending on the SYS_POWER_MANAGEMENT configuration.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

aurel32 commented:
> I have been playing with that PR, or rather the "soc: arm: st_stm32 add low power to stm32l4 series" commit for a few days on an STM32L422 based board. I haven't use the lptim related committed, and didn't notice any issue related with the timer. Note however that I have configured `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32768` and `CONFIG_STM32_LPTIM_CLOCK=32768`.
> 
> Overall I have been able to get it working, with a few limitations though:
> 
>     * The I2C device doesn't work anymore when `CONFIG_I2C_STM32_INTERRUPT=y`. The wait for the interrupt is done with K_FOREVER which causes the MCU to go into `SYS_POWER_STATE_SLEEP_3` / STOP2. The I2C device is not supported in that mode (at least for I2C1, it seems is not true for I2C3). I have workarounded that by wrapping the `i2c_transfer` calls with `sys_pm_ctrl_disable_state(SYS_POWER_STATE_SLEEP_3)` and `sys_pm_ctrl_enable_state(SYS_POWER_STATE_SLEEP_3)`. Strangely enough, I do not have the issue with the SPI device.
> 
>     * I am using HSI16 as the CPU clock, as it is the one configured after getting out of sleep. I understand that adding PLL support is complex, but it might be nice to also have MSI support as it the MCU support that directly. MSI allows higher clock speeds (although some changes are needed to the clock driver to allow that) and can work in PLL mode with the LSE.
> 
>     * I measured a power consumption of around 600µA which is higher than expected. I have other devices on the board that are supposed to be in low power mode and use less than 10µA. I have a spare PCB, I'll solder only the STM32L422 in the next days to check if the issue comes from the MCU or not.
> 
> 
> Also I haven't tested deep sleep, but from what I understand from the comments, some features are missing so it might be a good idea to move it to a separate PR.
> 
> With all that said, in order to progress on STM32L4 low power support, it might be a good idea to merge that feature without deep sleep and without making `SYS_POWER_MANAGEMENT=y` being the default.

A small update on this:

I haven't looked at the recent commits (I'll try to do it soon), but from what I understood they are just a rebase the LPTIM fixes. On my side I have been using that PR on a custom STM32L422 board measuring and transmitting LORA data every 5 minutes. It has now been running for 6 weeks without any issue. Now that the SPI pins configuration has been fixed, the whole board uses around 4µA when sleeping between two measurements/transmissions. It is running fine on a CR2450 battery.

All that to say that this PR looks the good approach to me. I think it's ready besides maybe the default options that could be discussed. I'll try to review it again in the next days, if all is fine, I'll just approve it.

FRASTM commented:
removing unused var wakeupdelay

aurel32 commented on subsys/power/policy/policy_residency_stm32.c:
```diff
@@ -0,0 +1,87 @@
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+#ifdef CONFIG_STM32_LPTIM_CLOCK_LSE
+#define LPTIM_CLOCK 32768
+#define LPTIM_TIMEBASE 0xFFFF
+#else
+#define LPTIM_CLOCK 32000
+#define LPTIM_TIMEBASE 0xF9FF
+#endif
+
+#define LPTIM_MAX_MS (((LPTIM_TIMEBASE + 1) / LPTIM_CLOCK) * 1000)
```
This is not used later.

FRASTM replied:
right, removed

aurel32 commented on subsys/power/policy/policy_residency_stm32.c:
```diff
@@ -0,0 +1,87 @@
+
+#define SECS_TO_TICKS		CONFIG_SYS_CLOCK_TICKS_PER_SEC
+
+u32_t table_prescaler[8] = {
+	LL_LPTIM_PRESCALER_DIV1,
+	LL_LPTIM_PRESCALER_DIV2,
+	LL_LPTIM_PRESCALER_DIV4,
+	LL_LPTIM_PRESCALER_DIV8,
+	LL_LPTIM_PRESCALER_DIV16,
+	LL_LPTIM_PRESCALER_DIV32,
+	LL_LPTIM_PRESCALER_DIV64,
+	LL_LPTIM_PRESCALER_DIV128,
+};
```
Same this is not used later.

FRASTM replied:
right, removed

aurel32 commented on subsys/power/policy/policy_residency_stm32.c:
```diff
@@ -0,0 +1,87 @@
+		if ((ticks == K_TICKS_FOREVER) ||
+		    (ticks >= pm_min_residency[i])) {
+			/* Set timeout for wakeup event */
+			if (ticks == K_TICKS_FOREVER) {
+				/* simply un-connect the LPTIM */
+				LL_APB1_GRP1_ForceReset(
+					LL_APB1_GRP1_PERIPH_LPTIM1);
+				LL_APB1_GRP1_DisableClock(
+					LL_APB1_GRP1_PERIPH_LPTIM1);
+			}
```
I guess those line is the whole purpose of the STM32 specific policy. Isn't that already taken into account in the LPTIM driver? There is the same check for `K_TICKS_FOREVER` and the same action of disabling the timer.

BTW what's the corresponding use case for that? I imagine waiting for another interrupt to wake-up the CPU, for example a GPIO interrupt from a user accessible button. In that case how is the timer re-enabled? It also probably means that the time is not counted during that long sleep, but I am not sure it's an issue.

FRASTM replied:
Yes canceling the lptim here is not required, as this is done in the lptim driver, removing.

FOREVER: if the system goes to sleep without a timer to wakeup, only a reset or EXTI can resume, if the pin has been configured before entering the sleep mode.

FRASTM commented:
including @aurel32 comments

FRASTM commented:
rebase after fix in the lptim

erwango commented on subsys/power/device.c:
```diff
@@ -45,6 +45,16 @@ static const char *const core_devices[] = {
+#define NUM_CORE_DEVICES	4
+#define MAX_DEV_NAME_LEN	16
+static const char core_devices[NUM_CORE_DEVICES][MAX_DEV_NAME_LEN] = {
+	"CLOCK_32K",
+	"CLOCK_16M",
+	"sys_clock",
+	"UART_0",
```
These are not STM32 devices, should this list left empty for now?

FRASTM replied:
changed to 
#define MAX_PM_DEVICES	1
static const char *const core_devices[] = {
	"",

aurel32 replied:
`"sys_clock"` was actually correct, it corresponds to file `drivers/timer/sys_clock_init.c` which maps to the LPTIM in that case.

FRASTM replied:
changing to :
```
#elif defined(CONFIG_SOC_SERIES_STM32L4X) || defined(CONFIG_SOC_SERIES_STM32WBX)
#define MAX_PM_DEVICES	2
static const char *const core_devices[] = {
	"sys_clock",
};
```

aurel32 reviewed (changes requested):
Beside the two changes I pointed, all looks good now. I am able to get an average sleeping power consumption of around 5µA for my whole board based on a STM32L422, including the energy needed to wake-up every 2 seconds to restart the timer.

FRASTM commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -7,6 +7,11 @@
+config STM32_LPTIM_TIMER
+	bool
+	default y
+	depends on DEVICE_POWER_MANAGEMENT
```
yes, I'm  just  correcting this

FRASTM commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -0,0 +1,135 @@
+	case SYS_POWER_STATE_SLEEP_3:
+#endif /* CONFIG_HAS_SYS_POWER_STATE_SLEEP_3 */
+		LL_LPM_DisableSleepOnExit();
+		if (IS_ENABLED(CONFIG_STM32_LPTIM_TIMER) &&
+		 LL_APB1_GRP1_IsEnabledClock(LL_APB1_GRP1_PERIPH_LPTIM1)) {
+			/* simply clock the LPTIM now */
+			LL_APB1_GRP1_EnableClock(
+				LL_APB1_GRP1_PERIPH_LPTIM1);
+		}
```
right

FRASTM commented:
stm32l4 low power modes depends on the CONFIG_SYS_POWER_MANAGEMENT

erwango commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -7,6 +7,11 @@
 
 if SOC_FAMILY_STM32
 
+config STM32_LPTIM_TIMER
```
Since this is not yet supported in all series (to prevent wrong usages), please move this in:
* soc/arm/st_stm32/stm32l4/Kconfig.defconfig.series
and
* soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series

Also we should remove equivalent config in boards configuration (disco_l475_iot1 and nucleo_wb5rg).

erwango reviewed (changes requested):
Last change

aurel32 reviewed (approved):
Thanks for the changes, it looks good to me.

FRASTM commented:
move the 
```
config STM32_LPTIM_TIMER
	bool
	default y
	depends on SYS_POWER_MANAGEMENT
```
from `soc/arm/st_stm32/common/Kconfig.defconfig.series` to soc/arm/st_stm32/stm32l4/Kconfig.defconfig.series

Arjun765 commented:
Is this applicable to WB55 as well? With the following code


```
#include <zephyr.h>
#include <sys/printk.h>

void main(void)
{
	
	while (true) {
		printk("Hello World! %s\n", CONFIG_BOARD);
		k_sleep(K_SECONDS(10));
	}
}
```

With config 
```
CONFIG_SYS_POWER_MANAGEMENT=y
CONFIG_STM32_LPTIM_TIMER=y
```

This consumes 2.8mA of current.

erwango commented:
@Arjun765, please note that this PR is almost one year old and impacts L4 only.
Though, PM has been extended to STM32WB series meantime. I'd advise to consider recent version on Zephyr as there were recent changes on PM subystem:
- For instance CONFIG_SYS_POWER_MANAGEMENT has been replaced by CONFIG_PM.
- PM support requires some device tree configuration.
